[WIP] Massive refactor

JakeGinnivan opened the pull request:
Getting some early visibility of this refactoring. It is now compiling and all the tests are switched to new API.

Will comments on some to callout the major changes. As I get tests passing things will likely change more.

Also, INotifyPropertyChanged support was dropped, will reintroduce once this is working

JakeGinnivan committed:
Trying to clean things up

JakeGinnivan committed:
Compiling again!

JakeGinnivan commented on src/ReactGraph/DependencyEngine.cs:
```diff
@@ -117,5 +98,272 @@ public void CheckCycles()
             throw new CycleDetectedException(sb.ToString().Trim());
         }
+
+        public void AddExpression<T>(ISourceDefinition<T> source, ITargetDefinition<T> target, Action<Exception> onError)
```
This is now the way nodes are added to the graph. Always in pairs. 

Once is a source the other a target, and an error handler has to be provided.

JakeGinnivan committed:
Moved a bunch of files out, and fixed up a few issues

JakeGinnivan committed:
Fixed issues after rebase

JakeGinnivan committed:
Fixing issues with expression parser

JakeGinnivan committed:
Lamdas removed from ExpressionStringBuilder

JakeGinnivan committed:
Pulled out expression adder which adds two definitions to the graph

JakeGinnivan committed:
Cleanup

JakeGinnivan committed:
Concept of root is wrong, an expression can have multiple roots (method call with params)

JakeGinnivan committed:
NodeInfo must override equals

JakeGinnivan committed:
Fixed issue with not being able to lookup nodes in the middle of a path, i.e Foo in a.Foo.Bar

JakeGinnivan committed:
Fixed cycle detection tests

JakeGinnivan committed:
Fixed a few more tests

JakeGinnivan commented:
@odeheurles this is getting closer now.

One of the biggest side effects is that roots are now part of the graph. i.e `viewModel.Foo` will have two nodes in the graph.

I was thinking we could tag then and they could be excluded from any visualisations.

odeheurles commented on src/ReactGraph.Tests/ExpressionParserTest.cs:
```diff
@@ -54,19 +47,18 @@ public void GetInvertedHarderNode()
-            validationErrorNode.Dependencies.Count.ShouldBe(1);
-            validationErrorNode.ShouldBeOfType<MemberDependencyDescriptor<bool>>()
-                .MemberInfo.ShouldBe(typeof(ScheduleViewModel).GetProperty("HasValidationError"));
+            var node = ExpressionParser.GetChildSources(expr);
```
rename the node variable?

odeheurles commented on src/ReactGraph.Tests/ExpressionParserTest.cs:
```diff
@@ -74,8 +66,38 @@ public void SimpleMethod()
             Expression<Func<int>> expr = () => Negate(simple.Value);
-            var node = expressionParser.GetFormulaInfo(expr);
-            node.Dependencies.Count.ShouldBe(1);
+            var node = ExpressionParser.GetChildSources(expr);
```
same here?

odeheurles commented on src/ReactGraph/BuilderBase.cs:
```diff
@@ -0,0 +1,50 @@
+            if (propertyInfo != null && !propertyInfo.CanWrite) return false;
+            if (fieldInfo != null && fieldInfo.IsInitOnly) return false;
+
+            return true;
```
shouldn't this return false by default and true only when it's actually writable?

odeheurles commented on src/ReactGraph/Construction/ExpressionStringBuilder.cs:
```diff
@@ -22,14 +22,20 @@ public static string ToString(Expression expression)
-            Out("(");
-            Out(String.Join(",", node.Parameters.Select(n => n.Name)));
-            Out(") => ");
+            if (node.Parameters.Any())
```
better

odeheurles commented on src/ReactGraph/Construction/ExpressionStringBuilder.cs:
```diff
@@ -22,14 +22,20 @@ public static string ToString(Expression expression)
             visitor.Visit(expression);
-            return visitor.builder.ToString();
+            var s = visitor.builder.ToString();
+            if (s.StartsWith("(") && s.EndsWith(")"))
```
are you sure?

"(A + B) \* (C - D)" will print "A + B) \* (C - D"
?

JakeGinnivan replied:
Good point, damn

JakeGinnivan commented:
Root nodes are now in the graph. Currently this makes visualisations go from:

![image](https://cloud.githubusercontent.com/assets/453152/4049244/69bf74d6-2d4b-11e4-918d-0858d4be59f7.png)

to

![image](https://cloud.githubusercontent.com/assets/453152/4049247/76ef22f0-2d4b-11e4-9fc8-848632e2a40f.png)

Is this an issue, or should I fix before merge

JakeGinnivan committed:
Approved changes to visualisation tests (is ordering, graph looks the same)

JakeGinnivan committed:
This looks right to me now, not sure why it was 8?

JakeGinnivan committed:
Only remove surrounding brackets if there are two brackets total in expression

AdaptiveBuild commented:
That's fine

Olivier

> On 26 Aug 2014, at 19:04, Jake Ginnivan notifications@github.com wrote:
> 
> Root nodes are now in the graph. Currently this makes visualisations go from:
> 
> to
> 
> Is this an issue, or should I fix before merge
> 
> —
> Reply to this email directly or view it on GitHub.

JakeGinnivan commented:
Rebased, removed the root nodes from visualisations and fixed your comment about the brackets.

Notify property changed not done yet, will see if I can get it done in the morning

JakeGinnivan commented:
I've fixed this :P

Sent from my Windows Phone

---

From: AdaptiveBuildmailto:notifications@github.com
Sent: ý26/ý08/ý2014 20:22
To: AdaptiveConsulting/ReactGraphmailto:ReactGraph@noreply.github.com
Cc: Jake Ginnivanmailto:jake@ginnivan.net
Subject: Re: [ReactGraph] [WIP] Massive refactor (#34)

That's fine

Olivier

> On 26 Aug 2014, at 19:04, Jake Ginnivan notifications@github.com wrote:
> 
> Root nodes are now in the graph. Currently this makes visualisations go from:
> 
> to
> 
> Is this an issue, or should I fix before merge
> 
> —
> Reply to this email directly or view it on GitHub.

—
Reply to this email directly or view it on GitHubhttps://github.com/AdaptiveConsulting/ReactGraph/pull/34#issuecomment-53467922.

JakeGinnivan committed:
Fixed some issues introduced with rebase

JakeGinnivan committed:
Fixed a few more tests

JakeGinnivan committed:
Step towards notify property changed

JakeGinnivan committed:
Hacked notify property changed support in, does not listen to leaf nodes yet, also reentrancy test is failing

JakeGinnivan commented:
Notify property changed support added. It is a hack, but works. Needs refactoring
